[Merged by Bors] - feat(miu_language): a formalisation of the MIU language described by D. Hofstadter in "Gödel, Escher, Bach".

gihanmarasingha opened the pull request:
We define an inductive type `derivable` so that `derivable x`  represents the notion that the MIU string `x` is derivable in the MIU language. We show `derivable x` is equivalent to `decstr x`, viz. the condition that `x` begins with an `M`, has no `M` in its tail, and for which `count I` is congruent to 1 or 2 modulo 3.

By showing `decidable_pred decstr`, we deduce that `derivable` is decidable.

---
<!-- put comments you want to keep out of the PR commit here -->

gihanmarasingha committed:
feat(miu_language): a formalisation of the MIU language described by D.
Hofstadter in "Gödel, Escher, Bach".

We show that a string `en` in the language is derivable if and only if
it satisfies `decstr en`, where `decstr` is a simple predicate that
'should' be decidable. However, we do not _prove_ decidability.

kim-em commented on archive/miu_language/basic.lean:
```diff
@@ -0,0 +1,249 @@
+... and list append.
+-/
+instance : has_append miustr :=
+⟨list.append⟩
```
How about:

```suggestion
/--
For simplicity, an `miustr` is just a list of `miu_atom`.
-/
@[derive [has_mem, has_append]]
def miustr := list miu_atom 
```

(perhaps with an additional comment saying what `derive` does?)

gihanmarasingha replied:
Thanks. I didn't know about @derive!

gihanmarasingha replied:
Still not sure what to do here. I managed to get
```lean
@[derive has_append]
def miustr := list miu_atom
```
to work, but the same technique didn't work for classes, such as `has_mem` with more than one parameter.

kim-em replied:
Ah, of course, sorry, I forgot about the other parameter of `has_mem`. Just leave that one as is (or use `by { dsimp [miustr], apply_instance }`).

gihanmarasingha replied:
Thanks. I'll just leave as is.

kim-em commented on archive/miu_language/basic.lean:
```diff
@@ -0,0 +1,249 @@
+⟨λ u, u.mrepr⟩ 
+
+/--
+In the other direction, we set up coercion from `string` to `miustr`.
```
```suggestion
In the other direction, we set up a coercion from `string` to `miustr`.
```

kim-em commented on archive/miu_language/basic.lean:
```diff
@@ -0,0 +1,249 @@
+1. x`I` → x`IU`,
+2. `M`x → `M`xx,
+3. x`III`y → x`U`y,
+4. x`UU`y → xy,
```
I think the backticks make this more confusing.

kim-em commented on archive/miu_language/basic.lean:
```diff
@@ -0,0 +1,249 @@
+Rule 4:  xUUy → xy
+
+For pedagogical purposes, we give definitons for the rules independently of the notion of
+derivability. We do not need these defintions to prove our main results. 
```
```suggestion
derivability. We do not need these definitions to prove our main results. 
```

kim-em replied:
I unconvinced that setting up the rules twice over makes for better pedagogy. Why not just do these examples after defining `derivable`?

gihanmarasingha replied:
OK. I've changed this as suggested. Good idea!

kim-em commented on archive/miu_language/decision_nec.lean:
```diff
@@ -0,0 +1,310 @@
+open miu_atom
+
+/-!
+### Numerical condition on the `I count
```
```suggestion
### Numerical condition on the `I` count
```

kim-em commented on archive/miu_language/decision_nec.lean:
```diff
@@ -0,0 +1,310 @@
+def icount : miustr → ℕ
+| [] := 0
+| (I::cs) := 1 + icount cs
+| (_::cs) := icount cs
```
Isn't there a built-in `count` function? Why not use that, and reuse a mathlib lemma about `count` and `append`?

gihanmarasingha replied:
I've made the suggested change. Much improved code follows. Thanks!

kim-em commented on archive/miu_language/decision_nec.lean:
```diff
@@ -0,0 +1,310 @@
+hk : derivable (z ++ repeat U (k * 2)) → derivable z,
+h : derivable (z ++ (repeat U (k * 2) ++ [U, U]))
+⊢ derivable (z ++ repeat U (k * 2))
+-/
```
What is this?

gihanmarasingha replied:
Woops. I must have hit the 'copy state to comment' button. Will remove.

kim-em commented on archive/miu_language/decision_nec.lean:
```diff
@@ -0,0 +1,310 @@
+`goodm xs` holds if `xs : miustr` begins with `M` and contains no `M` in its tail.
+-/
+def goodm (xs : miustr) : Prop :=
+  ∃ ys : miustr, xs = (M::ys) ∧ ¬(M ∈ ys)
```
Is it necessary to use an existential here? Can you just use `list.head` and `list.tail`?

gihanmarasingha replied:
I did that at first but ran into the issue that the empty list has a head.

kim-em replied:
I guess it is a bit gross to rely on the `default` element not being `M`. :-) Okay!

gihanmarasingha replied:
Hmm, I've taken on board this suggestion (with a bit of work), and I get `decidable_pred decstr` via the magic `@derive`. I still need to rewrite the `decision_suf.lean` file to work with the new definition.

Once done, I should be able to prove `decidable_pred derivable`. But it's 4am in the UK and I should get some sleep.

gihanmarasingha replied:
OK. I think I'm done now. Have made the change and thereby proved `decidable_pred derivable`.

kim-em commented on archive/miu_language/decision_nec.lean:
```diff
@@ -0,0 +1,310 @@
+
+open list
+
+lemma goodmrule1 (xs : miustr) (h₁ : derivable (xs ++ [I])) (h₂ : goodm (xs ++ [I]))
```
Here and elsewhere, we tend to use `snake_case` when naming lemmas.

gihanmarasingha replied:
Done.

kim-em commented on archive/miu_language/decision_nec.lean:
```diff
@@ -0,0 +1,310 @@
+/-
+We'll show, for each `i` from 1 to 4, that if `en` follows by rule `i` from `st` and if
+`goodm st` holds, then so does `goodm en`.
+-/
```
Comments like this should be `/-!` so they show up correctly in the typeset sources.

gihanmarasingha replied:
I think I've made the suggested changes, but don't have the document generator, so can't test.

kim-em commented on archive/miu_language/decision_nec.lean:
```diff
@@ -0,0 +1,310 @@
+-/
+
+theorem goodmder (en : miustr): derivable en →
+  goodm en:=
```
```suggestion
theorem goodmder (en : miustr): derivable en →
  goodm en :=
```

kim-em replied:
I don't understand the name here.

gihanmarasingha replied:
Fixed.

kim-em commented on archive/miu_language/decision_nec.lean:
```diff
@@ -0,0 +1,310 @@
+/--
+Suppose `en : miustr`. If `en` is `derivable`, then the condition `decstr en` holds.
+-/
+theorem decstr_of_der (en : miustr) : derivable en → decstr en:=
```
```suggestion
theorem decstr_of_der (en : miustr) : derivable en → decstr en :=
```

gihanmarasingha committed:
doc(miu_language): stylistic and typographical changes.

Co-authored-by: Scott Morrison <scott.morrison@gmail.com>

gihanmarasingha committed:
refactor(miu_language) using @derive to prove instances.

Co-authored-by: Scott Morrison <scott.morrison@gmail.com>

gihanmarasingha committed:
Revert "refactor(miu_language) using @derive to prove instances."

This reverts commit cc3fcb2173ec8037db66c3e6909852f325da5214.

gihanmarasingha committed:
style(miu_language): removed accidental extra comments.

gihanmarasingha committed:
style(miu_language) spelling. remove of backticks in some comments.

gihanmarasingha committed:
style(miu_language): removed duplicate rules of inference.

gihanmarasingha committed:
style(miu_language): minor changes to spacing, etc.

gihanmarasingha committed:
style(miu_language): fixed formatting for comments to be typeset.

gihanmarasingha committed:
refactor(miu_language): replace `icount`, `ucount` with mathlib `count`

gihanmarasingha committed:
style(miu_language): improved lemma-naming

gihanmarasingha committed:
refactor(miu_language): rewrite definition of `goodm`.

The new definition dispenses with `∃` and thereby facilitates automatic derivation of `decidable_pred decstr`.

gihanmarasingha committed:
refactor(miu_language): rewriting using new definition of `goodm`.

gihanmarasingha committed:
feat(miu_language): proved that `derivable` is an instance of `decidable_pred`

gihanmarasingha committed:
refactor(miu_language): simplified `goodm*` proofs.

gihanmarasingha committed:
refactor(miu_language): simplified auxiliary results needed for `der_of_decstr`.

kim-em commented on archive/miu_language/arithmetic.lean:
```diff
@@ -0,0 +1,101 @@
+    split;
+      norm_num,
+    refl,
+  }, { /- Induction step -/
```
I think we're going to want archive material to be an exemplar of "idiomatic Lean", including following the style guide.

Could you have a look at https://leanprover-community.github.io/contribute/style.html and try to follow it obsessively? The thing I noticed here was "braces on their own line".

gihanmarasingha replied:
I've updated the code. I _think_ it now adheres to the style guide. I take it that the variable naming convention is not inflexible - certainly the guide itself doesn't follow the variable naming convention.

Do let me know if you spot anything amiss. Thanks!

gihanmarasingha replied:
One other question: by refactoring, I've considerably reduced the size of `arithmetic.lean`. I now think I should remove this file and insert the lemmas from `arithmetic.lean` into the `decision_suf.lean` file where needed. Do you agree?

EDIT: `arithmetic.lean` already removed after review feedback from Johan Commelin.

kim-em commented on archive/miu_language/basic.lean:
```diff
@@ -0,0 +1,204 @@
+
+Additionally, we have an axiom
+
+* `MI` is derivable.
```
Presumably this is a distinction that Hofstadter makes, but it seems unnatural from an idiomatic point of view in a language with inductive types. More straightforward would be to say there are five constructors for an inductively defined predicate on strings. `MI` is only different in that it is a constructor without an argument.

I see that this is actually what you do below --- maybe you could restructure this part of the doc-strings to foreshadow that? i.e. explicitly explain that the inductive type with 5 constructors is more natural than making a distinction between "axioms" and "rules of inference". Or just discard the distinction entirely?

gihanmarasingha replied:
I hope I've made cleared this up with my latest commit. Please do let me know before I resolve the conversation. I'll continue working on the comments for `decision_suf.lean`.

jcommelin commented on archive/miu_language/arithmetic.lean:
```diff
@@ -0,0 +1,101 @@
+    /- The argument splits into two cases now, depending on whether
+    1 + 3(k+1) ≤ 2^g or not. If true, we take the new exponent to be g,
+    else we take it to be g+2. -/
+    by_cases hp : (1 + 3*nat.succ k≤ 2^g), { /- Two possibilities-/   
```
```suggestion
    by_cases hp : (1 + 3*(k+1) ≤ 2^g), { /- Two possibilities-/   
```

gihanmarasingha committed:
style(miu_language): better description of `derivable` and `derive`As per Scott Morrison's suggestion, I made it clearer that Hofstadter's 'axiom plus four rules of inference' becomes five constructors of an inductive type.  Further, I explained the use of `derive`.

gihanmarasingha committed:
refactor(miu_language): simplification of arithmetic result.

gihanmarasingha committed:
style(miu_language): indentiation, bracesAlso some minor refactoring, rewriting comments.

gihanmarasingha committed:
style(miu_language): minor spelling error

jcommelin commented on archive/miu_language/arithmetic.lean:
```diff
@@ -0,0 +1,47 @@
+open nat
+
+private lemma le_pow2_and_pow2_eq_mod3' (c : ℕ) (x : ℕ) (h : c = 1 ∨ c = 2) :
+  ∃ m : ℕ, c + 3*x ≤ 2^m ∧ 2^m % 3 = c % 3 :=
```
Do you want the `c % 3` at the end of the statement? You might just as well write `c`, since `c = 1` or `c = 2`.

gihanmarasingha replied:
True, but the current statement of this lemma makes unification in its corollary, `le_pow2_and_pow2_eq_mod3`, easier. I'd rather keep this as it is.

jcommelin replied:
Sure... fine with me.

jcommelin commented on archive/miu_language/arithmetic.lean:
```diff
@@ -0,0 +1,47 @@
+If `a` is 1 or 2 modulo 3, then exists `k` a power of 2 for which `a ≤ k` and `a ≡ k [MOD 3]`.
+-/
+lemma le_pow2_and_pow2_eq_mod3 (a : ℕ) (h : a % 3 = 1 ∨ a % 3 = 2) :
+  ∃ m : ℕ, a ≤ (pow 2 m) ∧ (pow 2 m) % 3 = a % 3:=
```
Why do you use `pow 2 m` instead of `2 ^ m`?

gihanmarasingha replied:
No good reason: I will change it to `2^m`.

jcommelin commented on archive/miu_language/arithmetic.lean:
```diff
@@ -0,0 +1,47 @@
+import tactic.linarith
+
+/-!
+# Arithmetic
```
I agree with your comment elsewhere that this file is now so short that it makes sense to make it a section in another file.
The lemmas don't seem directly useful outside of this project, but otherwise I would think they should actually move into the main `src/` tree of mathlib.

gihanmarasingha replied:
OK. I'll do that. The one lemma that _might_ be more generally useful is the fixed-point result:
```lean
lemma eq_of_countI_eq_length  {ys : miustr} (h : count I ys = length ys) :  ys = repeat I (count I ys) :=
```

Of course, this should be expressed in its generalisation:
```lean
variable α : Type*
variable [decidable_eq α]

lemma eq_of_count_eq_length {a : α} {l : list α} (h : count a l = length l) :  l = repeat a (count a l) :=
```

But I imagine that including every slightly useful lemma would cause mathlib to grow exponentially!

gihanmarasingha replied:
Other potentially useful lemmas (when suitably generalised) are:

```lean
lemma tail_append_singleton_of_ne_nil {a : miu_atom} {as : miustr} (h₁ : as ≠ nil) : 
  tail (as ++ [a]) = tail as ++ [a] :=
```
and
```lean
lemma exists_cons_of_ne_nil {as : miustr} (h : as ≠ nil) : ∃ c cs, as = c :: cs :=
```

kim-em replied:
I think all three of these are fine for the main `src/` directory!

jcommelin commented on archive/miu_language/basic.lean:
```diff
@@ -0,0 +1,215 @@
+
+The [MIU formal system](https://en.wikipedia.org/wiki/MU_puzzle) was introduced by Douglas
+Hofstadter in the first chapter of his 1979 book,
+[Gödel, Escher, Bach](https://en.wikipedia.org/wiki/G%C3%B6del,_Escher,_Bach).
```
Note that we have a `.bib` file in the repo. If you want you can add an entry there as well.

gihanmarasingha replied:
OK! I note there isn't an entry for TPIL. Is there a reason for that? It might be nice to refer to TPIL or another resource on inductive types.

kim-em replied:
It's certainly fine to add things to the bibliography! We should use it much more than we have been.

gihanmarasingha replied:
If I'm to edit `references.bib`, should I create a new PR or add the commit to this PR?

In the latter case, is it OK to `git merge master` into the miu branch or should I rebase (or something else)? Thanks!

bryangingechen replied:
I'd say you can just add the commit to this PR and it's better to merge master into this branch. Rebasing is generally discouraged after people start reviewing since it can screw up GitHub's review comments.

gihanmarasingha replied:
Thanks. I'm still finding my way around Git! I'm about to push the merged branch with my additions to `references.bib` and the corresponding addition of citations in `basic.lean`.

gihanmarasingha committed:
style(miu_language): renaming + moving lemmasMoved lemmas from `arithmetic.lean` to `decision_suf.lean` and removed `arithmetic.lean`. Renamed some lemmas. Added to documentation.

gihanmarasingha committed:
doc(miu_language): add tags

bryangingechen commented on archive/miu_language/decision_nec.lean:
```diff
@@ -0,0 +1,246 @@
+/-!
+### Numerical condition on the `I` count
+
+Suppose `st : miustr`. Then `count I st` is the number of `I`s in `st. We'll show, if
```
```suggestion
Suppose `st : miustr`. Then `count I st` is the number of `I`s in `st`. We'll show, if
```

(I found this using my [doc previewer](https://observablehq.com/@bryangingechen/github-lean-doc-preview?url=%223739%22#docs).)

gihanmarasingha committed:
doc(miu_language): added missing backtick.

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

gihanmarasingha committed:
Merge branch 'master' into miu

gihanmarasingha committed:
doc(refences.bib): add TPIL + Gödel, Escher, Bach entries

gihanmarasingha committed:
doc(miu_language): add references

gihanmarasingha committed:
feat(data/list/basic): add a result on repeat & count and two with premises `l ≠ nil`

Three results. The first states: if `count a = length l`, then `repeat a (count a l) = l`. The second states: if `l ≠ nil`, then `tail (l ++ [a]) = tail l ++ [a]`, if `l ≠ nil`. The last states: if `l ≠ nil`, then `∃ b L, l = b :: L`.

gihanmarasingha committed:
chore(miu_language): remove three lemmas

Remove three lemmas from `decision_nec` and `decision_suf`, having added them to `/src/data/list/basic.lean` in a previous commit.

robertylewis commented on archive/miu_language/basic.lean:
```diff
@@ -0,0 +1,223 @@
+/--
+We want to use list membership.
+-/
+instance : has_mem miu_atom miustr :=
```
This can be replaced by changing the `derive` on `miustr` to `@[derive [has_append, has_mem miu_atom]]`

robertylewis commented on archive/miu_language/decision_nec.lean:
```diff
@@ -0,0 +1,231 @@
+  intro h,
+  induction h,
+  { left, apply mod_def, },
+    any_goals {apply mod3_eq_1_or_mod3_eq_2 h_ih},
```
```suggestion
  any_goals {apply mod3_eq_1_or_mod3_eq_2 h_ih},
```

robertylewis commented on archive/miu_language/decision_nec.lean:
```diff
@@ -0,0 +1,231 @@
+begin
+  cases h₂ with mhead nmtail,
+  have k : as ≠ nil ,
+    { intro h, rw h at mhead, rw [nil_append] at mhead, contradiction, },
```
```suggestion
  { intro h, rw h at mhead, rw [nil_append] at mhead, contradiction, },
```
and below

robertylewis commented on archive/miu_language/decision_suf.lean:
```diff
@@ -0,0 +1,382 @@
+  use m,
+  split,
+  { convert hm.1, exact (mod_add_div a 3).symm, },
+  rw [hm.2, mod_mod _ 3],
```
```suggestion
  { rw [hm.2, mod_mod _ 3], }
```
I'm seeing this a lot in the PR. Personally I don't like the asymmetry; `split` creates two goals, so both should be in `{ }`, especially when both branches are of similar length/difficulty. But I'm not sure we have a written standard about this, do others agree?

gihanmarasingha replied:
I've made the changes you suggested, ensuring each remaining subgoal is in a brace-delimited indented block.

The format I used was one of the options suggested in TPIL (Section 5.4):
> [...] every time a tactic leaves more than one subgoal, we separate the remaining subgoals by enclosing them in blocks and indenting, until we are back down to one subgoal. Thus if the application of theorem foo to a single goal produces four subgoals, one would expect the proof to look like this:
> ```lean
> begin
>   apply foo,
>   { ... proof of first goal ... },
>   { ... proof of second goal ... },
>   { ... proof of third goal ... },
>   proof of final goal
> end
> ```

Do let me know if either (1) I've interpreted TPIL incorrectly or (2) the suggestions in TPIL are not sufficiently prescriptive for contributions to mathlib.

robertylewis replied:
No, you've read TPIL right! But I disagree with TPIL (except when the final goal is significantly longer than the first branch). And it looks like the [mathlib style guide](https://leanprover-community.github.io/contribute/style.html#tactic-mode) does too. We'll try to fix the discrepancy. Thanks for making the change.

Besides style, delimiting every subgoal is slightly more stable. If `foo` were changed so that `apply foo` created 5 goals instead of 4, the final proof could go haywire; the breakage is easier to diagnose when everything is focused. This is admittedly a minor point.

robertylewis commented on archive/miu_language/decision_suf.lean:
```diff
@@ -0,0 +1,382 @@
+  { refl, },
+  cases x,
+  { exfalso, exact hm (mem_cons_self M xs), }, -- case `x = M` gives a contradiction.
+  { rw [count_cons, if_pos (rfl),length,succ_eq_add_one,succ_inj'], -- case `x = I`
```
```suggestion
  { rw [count_cons, if_pos (rfl), length, succ_eq_add_one, succ_inj'], -- case `x = I`
```

robertylewis commented on archive/miu_language/decision_suf.lean:
```diff
@@ -0,0 +1,382 @@
+  { exfalso, rw count at h, contradiction, },
+  simp only [mem_cons_iff],
+  cases z,
+    repeat -- cases `z = M` and `z=I`
```
Is this multiple subgoals? `{}`?

gihanmarasingha replied:
I'm not certain I know how to format tactics that use `repeat`, `any_goals`, etc. I've made what I hope are acceptable revisions. Please check:

I have `any_goals` on on line 76 of `decision_nec.lean` and the above use of `repeat` now on line 312 of `decision_suf.lean`.

robertylewis replied:
I made an indentation suggestion here, otherwise it looks good. My rule of thumb: don't change the indentation for a tactic call except when entering a `{ ... }`. In that case, increase indentation by two.

robertylewis commented on archive/miu_language/decision_suf.lean:
```diff
@@ -0,0 +1,382 @@
+  rcases (exists_cons_of_ne_nil this) with ⟨z,zs,rfl⟩,
+  rw head at mhead,
+  rw mhead at *,
+  simp [count] at *,
```
Please try to avoid nonterminal calls to `simp`, here and below. `simp only` is fine.

robertylewis reviewed (commented):
This is looking nice!

gihanmarasingha committed:
style(miu_language): indentation, braces, etc.

* Use `derive` rather than explicit derivation of typeclass in `basic.lean`
* Ensure all remaining goals are presented in indented blocks.
* Eliminate nonterminal use of `simp`, replace with `simp only`.

gihanmarasingha committed:
style(miu_language): add spaces after commas

robertylewis commented on archive/miu_language/decision_suf.lean:
```diff
@@ -0,0 +1,393 @@
+    cases z,
+      repeat -- cases `z = M` and `z=I`
+      { right, apply hzs, simp only [count, countp, if_false] at h, rw ←h, refl, },
+      { left, refl, }, }, -- case `z = U`
```
```suggestion
    repeat -- cases `z = M` and `z=I`
    { right, apply hzs, simp only [count, countp, if_false] at h, rw ←h, refl, },
    { left, refl, }, }, -- case `z = U`
```

robertylewis committed:
Update archive/miu_language/decision_suf.lean

jcommelin commented on docs/references.bib:
```diff
@@ -327,3 +327,21 @@ @book{halmos1950measure
+          Soonho Kong},
+Title = {{T}heorem {P}roving in {L}ean},
+Year = {2017},
+howpublisehd = {\url{https://leanprover.github.io/theorem_proving_in_lean/}},
```
```suggestion
howpublished = {\url{https://leanprover.github.io/theorem_proving_in_lean/}},
```

gihanmarasingha replied:
Good catch!

jcommelin reviewed (commented):
Sometimes there are multiple empty lines following each other... I would just use 1 empty line in those cases. @robertylewis what do you think?
@gihanmarasingha Thank you so much for writing this! It looks great!

gihanmarasingha committed:
Update docs/references.bib

Co-authored-by: Johan Commelin <johan@commelin.net>

gihanmarasingha committed:
style(miu_language): removed duplicate new lines

gihanmarasingha committed:
Merge branch 'miu' of https://github.com/leanprover-community/mathlib into miu

jcommelin reviewed (approved):
Thanks :tada:

bors merge
